### ISSUE TITLE
[TypeScript] TransformCallback's Result argument as non-optional

### ISSUE THREAD

**Issue by aoberoi** [postcss/postcss#1213]
Hello there!

I noticed that the `TransformCallback` (return value from `postcss.plugin()`) in the type definitions receives an optional `result` argument.

https://github.com/postcss/postcss/blob/3a44f17aadb81f984e760d05cceac2e6e5b6471d/lib/postcss.d.ts#L36

I cannot think of a reason why `result` would be undefined. But leaving its type this way forces my plugin to deal with the possibility. I have to add an ugly block of code near the top of my transform callback that looks like this, just so that that I don't have to check for `undefined` later:

```javascript
    if (result === undefined) {
      throw new Error('Cannot run without a result defined');
    }
```

I propose that we change the definition to the following:

```typescript
interface TransformCallback {
    /**
     * @returns Asynchronous plugins should return a promise.
     */
     (root: Root, result: Result): Promise<any> | any;
  }
```

This should not break existing TypeScript implementations since a function which doesn't use the `result` argument would still implement this interface. You'll also notice that the return value is changed, which also shouldn't break anything, since both the existing return type and the one I've proposed uses a union with `any`, which is effectively still `any`. I just think this reads clearer, since asynchronous plugins should return a Promise, and otherwise the return value doesn't really matter.

If this sounds good to the maintainers, I'm happy to make the PR.

cc @jedmao (seems to be the resident typescript expert)

**Comment by jedmao** [comment]
You're right that a callback function shouldn't have optional args, as they are all passed regardless. You just opt into the ones you want by only asking for `n` number of args.

As for `void | Function | any`. Yeah, that's really just the same as `any`, but I wrote the union to be more explicit. If you look at the [`postcss.plugin`](http://api.postcss.org/postcss.html#.plugin) docs, you'll see two examples of return types: `void | Promise<void>`, but technically there's no reason you couldn't return other things too. Probably should have written this as `void | Promise<void>` without the `any` in the first place so people don't get the wrong idea. If they return anything else it's not like we're going to do anything with that information. And now here we are, wondering if anyone's relying on that return type for some reason? So it does have the potential to break some plugins if we remove `any` from that union, in that they won't compile w/o changing the return type to suit.

In conclusion, yes, we can make the `result` non-optional and `void | Promise<void> | any` for explicitness and backwards compatibility.

**Comment by aoberoi** [comment]
Thanks for the fast feedback @jedmao.

I actually think `void | Promise<void> | any` as the return type would be a little less clear. If I happen to return a `Promise<string>`, for example, it isn't assignable to `Promise<void>`. The only reason it would work is that its all caught by the `any` in the union. The same goes for return `number`, it wouldn't be assignable to `void`, but would continue to work because of the `any`.

Therefore I think `Promise<any> | any` is actually more accurate. It's like saying "I don't care, return `any`thing, I'm going to throw it away, unless its a `Promise`, and then I'm still going to throw away the resolved value."

**Comment by jedmao** [comment]
Is there some reason to return a string or a number of which I'm unaware? I think that's the point, that it's explicitly telling you that you can just write a function that doesn't return anything or you can return a promise that resolves with no value. Anything else is not how it was designed and I think the types should call out how it's designed and intended to be used.

I get you're point about "I'm going to throw it away" but `Promise<any>` doesn't actually tell you anything about throwing it away. You might wonder if PostCSS does something with the resolved or returned value, which is misleading. As such, I think `Promise<void>` is more clear, because we're really not expecting a value and, as you said, we're not using one if you provide it.

**Comment by aoberoi** [comment]
Hmm.. I guess the meaning of `any` in this situation is subjective. It’s totally valid that you feel it could mislead the developer into thinking the value has meaning, and yet it’s also totally valid that I think it’s more accurate and actually models the behavior better (I think `void` could signal that it’s important that there’s no value returned, hence TypeScript exhibiting the behavior that `number` is not assignable to `void`). 

I’d like you to consider an example, which is actually the situation for the plugin I’m writing. I return a function that can call itself. When it calls itself, it’s return value is important, and I chose that it returns a `Promise<Container>`. I use the `Container` value to recursively merge declarations back up to the `Root`. This isn’t really an edge case, as any plugin dealing with importing or processing nested structures would likely design them to do the same. From an implementation point of view, this works. However if the return type of `TransformCallback` was `Promise<void>`, I would need to do something like the following _just to satisfy the type system_:

```
postcss.plugin(‘foo’, (opts) => {
  return (root, result) => myRecursiveFunc(root, result).then(() => {});
});
```

I have to effectively erase the resolves value of my recursive function so that it’s assignable to `void`. That seems just as bad, if not worse, than the `throw` I’m using to convince the type system that `result` is never undefined.

This is a demonstration of the interpretation I referred to earlier: TypeScript says `Container` is not assignable to `void`, so I must interpret the return type of `void` as meaningful.

**Comment by jedmao** [comment]
Recursion is a great example of the need for a more loose type.
